gnosis-imap/gnosis-imap-spec.ini

```diff
@@ -148,6 +148,132 @@ proxy_storage:
 +requires:
     andBool 0 <Int DATA_LEN
 
+[removeOwner-success]
+k: (#execute => #halt) ~> _
+output: _ => _
+statusCode: _ => EVMC_SUCCESS
+callStack: _
+this: #PROXY_ID
+msg_sender: #PROXY_ID
+callData: #abiCallData("removeOwner", (#address(PREV_OWNER), 
+                                       #address(OWNER), 
+                                       #uint256(NEW_THRESHOLD)))
+callValue: 0
+wordStack: .WordStack => _
+localMem: .IMap => _
+pc: 0 => _
+gas: #gas(STARTGAS, 0, 0) => _
+memoryUsed: 0 => _
+log: _:List ( .List => ListItem(#abiEventLog(#PROXY_ID, "RemovedOwner", #address(OWNER))) )
+refund: _ => _
+coinbase: _ => _
+proxy_storage:
+    store(store(store(store(M1:IMap, #hashedLocation({COMPILER}, {OWNERS}, OWNER), INIT_OWNER),
+                                     #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER), INIT_PREV_OWNER),
+                                     #hashedLocation({COMPILER}, {THRESHOLD}, .IntList), THRESHOLD),
+                                     #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList), OWNERCOUNT)
+    => M2:IMap
++requires:
+    ; Address public constant SENTINEL_OWNERS = address(0x1);
+    andBool SENTINEL ==Int 1
+    ; Range
+    andBool #rangeAddress(  PREV_OWNER)
+    andBool #rangeAddress(  OWNER)
+    andBool #rangeAddress(  INIT_OWNER)
+    andBool #rangeAddress(  INIT_PREV_OWNER)
+    andBool #rangeUInt(256, NEW_THRESHOLD)
+    andBool #rangeUInt(256, THRESHOLD)  
+    andBool #rangeUInt(256, OWNERCOUNT)
+    andBool #rangeUInt(256, OWNERCOUNT -Int 1)
+    ; Requirements from code
+    ; Only allow to remove an owner, if threshold can still be reached.
+    andBool OWNERCOUNT -Int 1 >=Int NEW_THRESHOLD
+    ; Validate owner address and check that it corresponds to owner index
+    andBool OWNER =/=Int 0
+    andBool OWNER =/=Int SENTINEL
+    andBool INIT_PREV_OWNER ==Int OWNER
+    ; path condition
+    andBool OWNER =/=Int PREV_OWNER     // implied from 'owners[prevOwner] == owner' and 'owner != SENTINEL_OWNERS'
```

Comment:
@daejunpark I was also wondering why it worked and though about it. Initially I added the no-hash-collision requirement, but the spec (with just the ensure `select(M2, #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))    ==Int INIT_OWNER`) was still failing; therefore I added the requirement `OWNER =/=Int PREV_OWNER` and the spec was proved. Just by curiosity I dropped the no-hash-collision requirement and the spec was still proved. I was happy that the condition `OWNER =/=Int PREV_OWNER` is a consequence of the requirements from the Solidity code and no other cases are generated because of this.
I noticed that in Jenkins the spec is proved. How do you think I should handle this?